Jellyfin.Server.Implementations/Devices/DeviceManager.cs

```diff
@@ -0,0 +1,232 @@
+            return device;
+        }
+
+        /// <inheritdoc />
+        public async Task<DeviceOptions?> GetDeviceOptions(string deviceId)
+        {
+            await using var dbContext = _dbProvider.CreateContext();
+            return await dbContext.DeviceOptions
+                .AsQueryable()
+                .FirstOrDefaultAsync(d => d.DeviceId == deviceId)
+                .ConfigureAwait(false);
+        }
+
+        /// <inheritdoc />
+        public ClientCapabilities GetCapabilities(string deviceId)
+        {
+            return _capabilitiesMap.TryGetValue(deviceId, out ClientCapabilities? result)
+                ? result
+                : new ClientCapabilities();
+        }
+
+        /// <inheritdoc />
+        public async Task<DeviceInfo?> GetDevice(string id)
+        {
+            await using var dbContext = _dbProvider.CreateContext();
+            var device = await dbContext.Devices
+                .AsQueryable()
+                .Where(d => d.DeviceId == id)
+                .OrderByDescending(d => d.DateLastActivity)
+                .Include(d => d.User)
+                .FirstOrDefaultAsync()
+                .ConfigureAwait(false);
+
+            var deviceInfo = device == null ? null : ToDeviceInfo(device);
+
+            return deviceInfo;
+        }
+
+        /// <inheritdoc />
+        public async Task<QueryResult<Device>> GetDevices(DeviceQuery query)
+        {
+            await using var dbContext = _dbProvider.CreateContext();
+
+            var devices = dbContext.Devices.AsQueryable();
+
+            if (query.UserId.HasValue)
+            {
+                devices = devices.Where(device => device.UserId == query.UserId.Value);
+            }
+
+            if (query.DeviceId != null)
+            {
+                devices = devices.Where(device => device.DeviceId == query.DeviceId);
+            }
+
+            if (query.AccessToken != null)
+            {
+                devices = devices.Where(device => device.AccessToken == query.AccessToken);
+            }
+
+            if (query.Skip.HasValue)
+            {
+                devices = devices.Skip(query.Skip.Value);
+            }
+
+            var count = await devices.CountAsync().ConfigureAwait(false);
+
+            if (query.Limit.HasValue)
+            {
+                devices = devices.Take(query.Limit.Value);
+            }
+
+            return new QueryResult<Device>
+            {
+                Items = await devices.ToListAsync().ConfigureAwait(false),
+                StartIndex = query.Skip ?? 0,
+                TotalRecordCount = count
+            };
+        }
+
+        /// <inheritdoc />
+        public async Task<QueryResult<DeviceInfo>> GetDeviceInfos(DeviceQuery query)
+        {
+            var devices = await GetDevices(query).ConfigureAwait(false);
+
+            return new QueryResult<DeviceInfo>
+            {
+                Items = devices.Items.Select(ToDeviceInfo).ToList(),
```

Comment:
According to @cvium using this pattern is slower then expanding the LINQ
```suggestion
                Items = devices.Items.Select(i => ToDeviceInfo(i)).ToList(),
```

Reply:
I don't think it matters when the LINQ is being compiled into SQL by EF Core, unless @cvium has some information/data showing otherwise

Reply:
For hot code paths it is better to use lambdas over method groups because the method group is an extra allocation.

Reply:
Is that true when the linq is being compiled to SQL by EF Core? Regardless, this isn't a very hot path, so I don't think it matters much either way.

Reply:
It's true no matter what LINQ does. It's not really related to LINQ. You'd have to ask @Bond-009 what the official stance is.